routing/index_graph_starter.hpp

```diff
@@ -21,136 +26,180 @@ class IndexGraphStarter final
+      : m_junctionFrom(junctionFrom), m_junctionTo(junctionTo), m_type(type)
+    {
+    }
+
+    FakeVertex(FakeVertex const &) = default;
+    FakeVertex() = default;
+
+    bool operator==(FakeVertex const & rhs) const
+    {
+      return m_junctionFrom == rhs.m_junctionFrom && m_junctionTo == rhs.m_junctionTo &&
+             m_type == rhs.m_type;
+    }
+
+    Type GetType() const { return m_type; }
+
+    Junction const & GetJunctionFrom() const { return m_junctionFrom; }
+    m2::PointD const & GetPointFrom() const { return m_junctionFrom.GetPoint(); }
+    Junction const & GetJunctionTo() const { return m_junctionTo; }
+    m2::PointD const & GetPointTo() const { return m_junctionTo.GetPoint(); }
+
+  private:
+    Junction m_junctionFrom;
+    Junction m_junctionTo;
+    Type m_type = Type::PureFake;
+  };
+
+  struct FakeGraph
+  {
+    std::map<Segment, std::set<Segment>> m_outgoing;
+    std::map<Segment, std::set<Segment>> m_ingoing;
+    std::map<Segment, FakeVertex> m_segmentToVertex;
+    std::map<Segment, Segment> m_fakeToReal;
+    std::map<Segment, std::set<Segment>> m_realToFake;
+  };
+
+  void AddFakeVertex(Segment const & existentSegment, Segment const & newSegment,
+                     FakeVertex const & newVertex, bool isOutgoing, bool isPartOfReal,
+                     Segment const & realSegment);
+  void AddEnding(FakeEnding const & thisEnding, FakeEnding const & otherEnding, bool isStart,
+                 bool strictForward, uint32_t & fakeNumerationStart);
+  void AddStart(FakeEnding const & startEnding, FakeEnding const & finishEnding,
+                bool strictForward, uint32_t & fakeNumerationStart);
+  void AddFinish(FakeEnding const & finishEnding, FakeEnding const & startEnding,
+                 uint32_t & fakeNumerationStart);
+
+  // Returns existent segment if possible. If segment does not exist makes new segment,
+  // increments newNumber.
+  Segment GetSegment(FakeVertex const & vertex, uint32_t & newNumber) const;
+
+  void AddFakeEdges(vector<SegmentEdge> & edges) const;
+  void AddRealEdges(Segment const & segment, bool isOutgoing, std::vector<SegmentEdge> & edges) const;
   /// \brief If |isOutgoing| == true fills |edges| with SegmentEdge(s) which connects
   /// |fakeVertex| with all exits of mwm.
   /// \brief If |isOutgoing| == false fills |edges| with SegmentEdge(s) which connects
   /// all enters to mwm with |fakeVertex|.
   void ConnectLeapToTransitions(Segment const & segment, bool isOutgoing,
-                                vector<SegmentEdge> & edges);
+                                std::vector<SegmentEdge> & edges) const;
 
   WorldGraph & m_graph;
-  FakeVertex const m_start;
-  FakeVertex const m_finish;
+  uint32_t m_startId;
+  uint32_t m_finishId;
+  FakeGraph m_fake;
+};
+
+class FakeEdgesContainer final
+{
+  friend class IndexGraphStarter;
+
+public:
+  FakeEdgesContainer(IndexGraphStarter && starter)
```

Comment:
```c++
template <typename FG>
FakeEdgesContainer(uint32_t finishId, FG && fg): m_finishId(finishId), m_fake(forward<FG>(fg)) {}
```
I guess you don't need so strong connection between FakeEdgesContainer and IndexGraphStarter.

Reply:
How about to move FakeEdgesContainer into individual file?

Reply:
I want to hide inner structure of IndexGraphStarter from IndexRouter, IndexRoadGraph etc.
I think it's good to keep FakeVertex and FakeGraph private and do not have any public methods which allow to get raw m_fake from IndexGraphStarter. So FakeEdgesContainer must be a friend of IndexGraphStarter.

Reply:
In any case, please, move it to an individual file.

Reply:
moved